File: packages/maplibre_android/android/src/main/kotlin/com/github/josxha/maplibre/HostScopedRequestHeaders.java
package com.github.josxha.maplibre;

import androidx.annotation.Keep;
import java.util.Collections;
import java.util.HashMap;
import java.util.Locale;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

/**
 * Host-scoped request headers read by the MapLibre OkHttp interceptor.
 *
 * <p>Static methods are the jnigen entry points used to replace or clear headers for one exact
 * host.
 */
@Keep
public final class HostScopedRequestHeaders {
    private static final ConcurrentHashMap<String, Map<String, String>> headersByHost =
            new ConcurrentHashMap<>();

    private HostScopedRequestHeaders() {}

    public static void replace(String host, Map<String, String> headers) {
        final String normalizedHost = host.toLowerCase(Locale.ROOT);
        if (headers.isEmpty()) {
            headersByHost.remove(normalizedHost);
            return;
        }
        headersByHost.put(
                normalizedHost, Collections.unmodifiableMap(new HashMap<>(headers)));
    }

    public static void clear(String host) {
        headersByHost.remove(host.toLowerCase(Locale.ROOT));
    }

    public static Map<String, String> headersForHost(String host) {
        return headersByHost.get(host.toLowerCase(Locale.ROOT));
    }
}

File: packages/maplibre_android/android/src/main/kotlin/com/github/josxha/maplibre/HostScopedRequestHeaders.kt
package com.github.josxha.maplibre

import okhttp3.Dispatcher
import okhttp3.Interceptor
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import org.maplibre.android.module.http.HttpRequestUtil
import java.util.concurrent.atomic.AtomicBoolean

internal fun applyHostScopedRequestHeaders(request: Request): Request {
    val headers = HostScopedRequestHeaders.headersForHost(request.url.host)
    val managedNames = HostScopedRequestHeaders.configuredHeaderNames()
    if (headers == null && managedNames.none { request.header(it) != null }) {
        return request
    }
    val builder = request.newBuilder()
    for (name in managedNames) {
        if (headers == null || !headers.containsKey(name)) {
            builder.removeHeader(name)
        }
    }
    headers?.forEach { (name, value) -> builder.header(name, value) }
    return builder.build()
}

internal object HostScopedRequestHeadersInterceptor : Interceptor {
    override fun intercept(chain: Interceptor.Chain): Response = chain.proceed(applyHostScopedRequestHeaders(chain.request()))
}

internal fun installHostScopedRequestHeadersInterceptor() {
    if (!requestHeadersInterceptorInstalled.compareAndSet(false, true)) return
    val dispatcher = Dispatcher().apply { maxRequestsPerHost = 20 }
    val client =
        OkHttpClient
            .Builder()
            .dispatcher(dispatcher)
            // Network interceptors run on each redirect hop, so a header for one
            // host is removed when OkHttp follows a redirect to another host.
            .addNetworkInterceptor(HostScopedRequestHeadersInterceptor)
            .build()
    HttpRequestUtil.setOkHttpClient(client)
}

private val requestHeadersInterceptorInstalled = AtomicBoolean()
